File: webmap/src/module/BlockInfo.ts
export class BlockInfo {
    private readonly _data: Uint8Array;

    constructor(data: Uint8Array) {
        this._data = data;
    }

    get minY(): number {
        return this.getInt(8);
    }

    getBlock(index: number): Block {
        return new Block(this.getInt(12 + index * 4), this.minY);
    }

    private getInt(position: number): number {
        let val: number = 0;
        for (let i: number = 0; i < 4; i++) {
            val |= (this._data[position + i] & 0xFF) << (i * 8);
        }
        return val;
    }
}

export class Block {
    private readonly _block: number;
    private readonly _biome: number;
    private readonly _yPos: number;
    private readonly _minY: number;

    constructor(packed: number, minY: number) {
        this._block = packed >>> 22;
        this._biome = (packed << 10) >> 22;
        this._yPos = ((packed << 20) >> 20);
        this._minY = minY;
    }

    get block(): number {
        return this._block;
    }

    get biome(): number {
        return this._biome;
    }

    get yPos(): number {
        return this._yPos + this._minY;
    }
}

export type Palette = {
    index: number;
    block: string;
}

File: plugin/src/main/java/net/pl3x/map/render/task/builtin/BlockInfo.java
package net.pl3x.map.render.task.builtin;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.file.Path;
import net.minecraft.core.BlockPos;
import net.minecraft.world.level.biome.Biome;
import net.minecraft.world.level.block.Block;
import net.pl3x.map.PaletteManager;
import net.pl3x.map.render.image.Image;
import net.pl3x.map.render.job.iterator.coordinate.RegionCoordinate;
import net.pl3x.map.render.task.Renderer;
import net.pl3x.map.render.task.ScanData;
import net.pl3x.map.render.task.ScanTask;
import net.pl3x.map.util.ByteUtil;
import net.pl3x.map.util.FileUtil;
import net.pl3x.map.util.Palette;

public class BlockInfo extends Renderer {
    private ByteBuffer byteBuffer;

    public BlockInfo(String name, ScanTask scanTask) {
        super(name, scanTask);
    }

    @Override
    public void allocateData() {
        this.byteBuffer = ByteBuffer.allocate(Image.SIZE * Image.SIZE * 4 + 20);
    }

    @Override
    public void saveData() {
        Path tilesDir = getScanTask().getWorld().getWorldTilesDir();
        // TODO - make work for higher zoom levels?
        Path dir = tilesDir.resolve(String.format(Image.DIR_PATH, 0, getName()));
        String filename = String.format(Image.FILE_PATH, getRegion().getRegionX(), getRegion().getRegionZ(), "pl3xmap.gz");
        try {
            FileUtil.saveGzip(this.byteBuffer.array(), dir.resolve(filename));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

    }

    @Override
    public void scanData(RegionCoordinate region, ScanData.Data scanData) {
        int minY = getWorld().getLevel().getMinBuildHeight();

        this.byteBuffer.clear();

        this.byteBuffer.put(0, ByteUtil.toBytes(0x706C3378)); // pl3x
        this.byteBuffer.put(4, ByteUtil.toBytes(0x6D617001)); // map1
        this.byteBuffer.put(8, ByteUtil.toBytes(region.getRegionX()));
        this.byteBuffer.put(12, ByteUtil.toBytes(region.getRegionZ()));
        this.byteBuffer.put(16, ByteUtil.toBytes(minY));

        Palette<Block> blockPalette = PaletteManager.INSTANCE.getBlockPalette();
        Palette<Biome> biomePalette = PaletteManager.INSTANCE.getBiomePalette(getWorld());

        for (ScanData data : scanData.values()) {
            boolean fluid = data.getFluidPos() != null;

            Block block = (fluid ? data.getFluidState() : data.getBlockState()).getBlock();
            Biome biome = fluid ? data.getFluidBiome() : data.getBlockBiome();
            BlockPos pos = fluid ? data.getFluidPos() : data.getBlockPos();

            int blockIndex = blockPalette.get(block).getIndex();
            int biomeIndex = biomePalette.get(biome).getIndex();
            int yPos = pos.getY() - minY; // ensure bottom starts at 0

            // 11111111111111111111111111111111 - 32 bits - (4294967295)
            // 1111111111                       - 10 bits - block (1023)
            //           1111111111             - 10 bits - biome (1023)
            //                     111111111111 - 12 bits - yPos  (4095)
            int packed = ((blockIndex & 1023) << 22) | ((biomeIndex & 1023) << 12) | (yPos & 4095);
            int index = (pos.getZ() & Image.SIZE - 1) * Image.SIZE + (pos.getX() & Image.SIZE - 1);
            this.byteBuffer.put(20 + index * 4, ByteUtil.toBytes(packed));
        }
    }
}
